sectionproperties/pre/sections.py

```diff
@@ -1,259 +1,568 @@
+        to a new absolute location. Returns a new Geometry representing the original 
+        with the selected point(s) shifted to the new location.
+
+        Points are identified by their index, their relative location within the points
+        list found in self.points. You can call self.plot_geometry(labels="points") to
+        see a plot with the points labeled to find the appropriate point indexes.
+
+        :param point_idxs: An integer representing an index location or a list of integer
+        index locations.
+        :type point_idxs: Union[int, List[int]]
+        :param dx: The number of units in the x-direction to shift the point(s) by
+        :type dx: float
+        :param dy: The number of units in the y-direction to shift the point(s) by
+        :type dy: float
+        :param abs_x: Absolute x-coordinate in coordinate system to shift the 
+        point(s) to. If abs_x is provided, dx is ignored. If providing a list
+        to point_idxs, all points will be moved to this absolute location.
+        :type abs_x: Optional[float]
+        :param abs_y: Absolute y-coordinate in coordinate system to shift the 
+        point(s) to. If abs_y is provided, dy is ignored. If providing a list
+        to point_idxs, all points will be moved to this absolute location.
+        :type abs_y: Optional[float]
+
+        :return: Geometry object with selected points translated to the new location.
+        :rtype: :class:`sections.pre.sections.Geometry`
+
+        The following example expands the sides of a rectangle, one point at a time,
+        to make it a square.
 
-        self = pre.GeometryCleaner(self, verbose).clean_geometry()
+            import sectionproperties.pre.sections as sections
+
+            geometry = sections.rectangular_section(d=200, b=150)
+            one_pt_shifted_geom = geometry.shift_points(point_idxs=1, dx=50) # Using relative shifting
+            both_pts_shift_geom = one_pt_shift_geom.shift_points(point_idxs=2, abs_x=200) # Using absolute relocation
+        """
+        current_points = copy.copy(self.points)
+        current_facets = copy.copy(self.facets)
+        current_holes = copy.copy(self.holes)
+        current_control_points = copy.copy(self.control_points)
+
+        if isinstance(point_idxs, int):
+            point_idxs = [point_idxs]
+        new_x, new_y = None, None
+        for point_idx in point_idxs:
+            current_x, current_y = current_points[point_idx]
+            new_x = abs_x if abs_x else current_x + dx
+            new_y = abs_y if abs_y else current_y + dy
+            current_points[point_idx] = (new_x, new_y)
+
+        new_geom = Geometry.from_points(
+            current_points, current_facets, current_holes, current_control_points,
+        )
+        return new_geom
 
-    def plot_geometry(self, ax=None, pause=True, labels=False, perimeter=False):
+    def plot_geometry(
```

Comment:
perimeter = True does not seem to trigger the linewidth = 3 line below?

Reply:
This is related to the calculation of the exterior perimeter. If the shapes are not disjoint, I can calculate (and plot) the exterior perimeter. If shapes are disjoint, then `CompoundGeometry.calculate_perimeter()` returns -1.0. In the instance of shapes being disjoint, what are your thoughts on the behaviour of `CompoundGeometry.plot_geometry(perimeter=True)`?

Reply:
If disjoint I'd love to see it reflected on the plot somehow as a warning to the user, for example a red perimeter around the disjoint object(s) as you'd have several external perimeters?

Reply:
I wonder since the results are invalid if you have disjoint geometry if the code should simply raise an exception and exit? Currently you can get the results either way, which if you don't realise by plotting geometry could lead to incorrect results. Perhaps you could have an optional option to halt_on_disjoint_geometry or similar as a thought?

Reply:
I don't think the results are invalid for disjoint geometry. I just did a test against a handcalc for two separate but equal rectangles and the result is correct (see below). Tell me more about this experience you had with incorrect results. 

To be honest, I _could_ increase the lineweight for disjoint geometry, too. It's just more leg-work because it will require at least three or four helper functions to do (iterate over Geometry in unionized CompoundGeometry, accumulate exterior rings, create facet lists for each exterior ring, and pass those in sequence to `.plot_geometry()`) and I am being a baby about doing it right now :)

![image](https://user-images.githubusercontent.com/38441489/115498937-52b3c400-a223-11eb-9596-52c6992e5439.png)

Reply:
I should have said some of the properties are invalid. Robbie could probably add more to the reasons, but torsion and warping functions come to mind. It was in the docs previously for example that the results are only valid if you have no disjoint geometry. It's also come up in standard responses from Robbie in some prior issues. 

Understood on highlighting disjoint geometry. To be honest shapely version is much less error prone in this respect. I've had instance where I've put elements in contact but forgot to clean_geometry previously and they haven't meshed properly. At least with shapely version this is no longer an issue are the merging/cleaning is all handled by shapely seemlessly.

Reply:
Ah! Thank you for the details. That all makes sense.

Interestingly, if you create two shapes and put them in contact as a CompoundGeometry, shapely will turn the svg image red because they are invalid for shapely (but valid for sectionproperties). If they stay green, then they are not touching. So there is something like this built in to the new preprocessor.

Regarding the perimeter arg in plot_geometry(), I will perhaps remove it as an argument from CompundGeometry for now until that is implemented.

Reply:
I should add that invalid geometry in shapely only affects certain shapely operations which will raise an error when given invalid geometry to operate on. We have not been using such operations in sectionproperties.

Reply:
> Regarding the perimeter arg in plot_geometry(), I will perhaps remove it as an argument from CompundGeometry for now until that is implemented.

Yeah makes sense.